### PR TITLE
H5MD format rewrite of Zarrtraj

### LINKED ISSUE #26: Tests failing due to misuse of "rmdir"
Modify aws tests to remove the group at the specified string, rather than attempting to call rmdir on the string.

### PR DESCRIPTION
**Pull request by ljwoods2**
Fixes #26, #27, #33 

Changes made in this Pull Request:
 - Complete rewrite of zarrtraj
 - Kerchunk patched temporarily before zarr 3.0 is released OR kerchunk supports links
 - Tests should run faster now- all s3 tests run through one server and one bucket

Won't delete old zarrtraj files until PR is ready

## Handling of H5MD flexibility

A ton has changed in this PR so I'll summarize how I'm handling H5MD flexibility here:

1. Though H5MD says time data is optional, this could lead to problems in the MDAnalysis frame-based model- say an H5MD element is the only element which contains data at a certain integration step. Then MDAnalysis would be forced to produce a timestep without time data. This time can't be inferred since you could have

positions/step: [1, 2, 3]
positions/time: [1.0, 2.0, 3.0]
velocities/step: [1, 2, 3, 100]

Therefore, I think the best thing to do is to just require that all time-dependent data has fixed or explicit time.

2. Links don't yet exist in Zarr, but in H5MD, it is required that positions/step is a hard link to box/edges/step. I handle this by just making sure the step datasets are all equal. In the kerchunk patch, linked datasets are simply duplicated, so this is the closest to the intended spec I can get on this requirement.

3. Since there are both "global" and "per-particle group" observables, the "per-particle group" ones are loaded into ts.data as "<group_name>/<dataset_name>" and global ones are loaded as "<dataset_name>" to prevent name mangling

4. In frames where time-dependent observables don't have any data, the ts.data key for that observable is deleted

5. ALL datasets that have data of a certain kind (i.e. box and positions both have "length" data) have to have the same unit. You can't have "positions" and "velocities" with different time units though this may be allowed in H5MD.

6. Though time-independent positions, velocities, and forces are allowed in H5MD, zarrtraj will fail out.

7. If there is only one group in "/particles", zarrtraj will select it to read. However, if there is more than one, the "group" kwarg must be provided (since H5MD allows any number of groups) else zarrtraj will fail out rather than assume which group the user wanted.

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: zarrtraj/ZARR.py [hunk 1]
```diff
@@ -0,0 +1,1567 @@
...
+    def n_frames(self):
+        """number of frames in trajectory"""
+        return len(self._global_steparray)
+
+    @staticmethod
+    def _format_hint(thing):
+        """Can this Reader read *thing*"""
+        # nb, filename strings can still get passed through if
+        # format='H5MD' is used
+        pass
```
Comment (hmacdope) [inline zarrtraj/ZARR.py:714]: I  assume we need this?

File: zarrtraj/data/create_COORDINATES_MISSING_H5MD_GROUP.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def main():
+    z = zarr.open_group(COORDINATES_SYNTHETIC_ZARRMD, mode="r")
+    outz = zarr.open_group(
+        "COORDINATES_MISSING_H5MD_GROUP_ZARRMD.zarrmd", mode="a"
+    )
+    zarr.convenience.copy_all(z, outz)
+    create_COORDINATES_MISSING_H5MD_GROUP(outz)
+
```
Comment (hmacdope) [inline zarrtraj/data/create_COORDINATES_MISSING_H5MD_GROUP.py:23]: You can make these and the other scripts testing fixtures.

**Review by hmacdope — APPROVED** [review]
Only nitpicks mostly and some gentle encouragement to clean up some cruft. But overall this is looking fantastic. I say go ahead and merge and work from there. We can clean up later. 

I also agree with the small liberties taken with the format. We can always re-asses if we come across files that break this.

File: zarrtraj/data/create_COORDINATES_MISSING_H5MD_GROUP.py (hunk 2 above)
Comment (hmacdope) [reply zarrtraj/data/create_COORDINATES_MISSING_H5MD_GROUP.py:23]: If its easier for you, that way you can construct the h5md "file" in place.

File: zarrtraj/ZARR.py [hunk 3]
```diff
@@ -0,0 +1,1567 @@
...
+        does *not* close and open the file like most readers because
+        the information about the MPI communicator would be lost; instead
+        it rewinds the trajectory back to the first timstep.
+
+        """
+        self.close()
+        self._open_trajectory()
+
+    def Writer(self, filename, n_atoms=None, **kwargs):
+        """Return writer for trajectory format
```
[the remark this answers is not in the document]
Comment (ljwoods2) [reply zarrtraj/ZARR.py:543]: Tons of the docstrings are just copied from H5MD in the core codebase, haven't even touched them. I'll make sure to update all of these once I'm more confident I won't make massive changes

File: zarrtraj/ZARR.py (hunk 1 above)
Comment (ljwoods2) [reply zarrtraj/ZARR.py:714]: Yes, this is actually something I want your thoughts on- should this reader take precedence over the core H5MDReader for reading H5MD files from disk if the user imports it? If so, I can just catch all files with the .zarrmd or .h5md file extension and don't need a format hint. If not, when and how should the user specify they want to use this reader to read their h5md file?

File: zarrtraj/ZARR.py [hunk 4]
```diff
@@ -0,0 +1,1567 @@
...
+                return file["particles"][group][dset]["value"].shape[1]
+        raise NoDataError(
+            "Could not construct minimal topology from the "
+            "H5MD trajectory file, as it did not contain a "
+            "'position', 'velocity', or 'force' group. "
+            "You must include a topology file."
+        )
+
+
+class H5MDWriter(base.WriterBase):
```
[the remark this answers is not in the document]
Comment (ljwoods2) [reply zarrtraj/ZARR.py:745]: Yes, it is just copied for now to have something to look at while I write the zarr writer, will delete this later

**Pull request merged**